### PR TITLE
Add fuzzing setup for `Message::decode`

### PR DESCRIPTION
**Pull request by wiktor-k**
See `fuzz/README.md` for details on how to run it.

### CODE REVIEW

**Comment by wiktor-k** [comment]
@kpcyrd Sorry for disturbing you but if you don't mind could you look at this fuzzer setup and see if anything is obviously broken?

Thanks in advance! :bow: and have a nice day! :wave:

File: fuzz/Cargo.lock
Comment (kpcyrd) [inline fuzz/Cargo.lock:1]: I'd recommend to avoid having a second `Cargo.lock` and instead make it a workspace member by adding this to the parent `Cargo.toml`:

```
[workspace]
members = ["fuzz"]
```

This was recently made possible in cargo-fuzz 0.11.4 (https://github.com/rust-fuzz/cargo-fuzz/pull/357) :)

File: fuzz/fuzz_targets/message_decode.rs
Comment (kpcyrd) [inline fuzz/fuzz_targets/message_decode.rs:9]: it likely won't have any effect on the compiled code, but you could write this as:

```suggestion
    let _ = Message::decode(&mut &data[..]);
```

or by writing `|mut data: &[u8]| {`

Comment (wiktor-k) [reply fuzz/fuzz_targets/message_decode.rs]: I've used your suggestion and squashed the commit (with your name in `Co-authored-by` trailer). the `|mut data..|` didn't seem to work due to the issue with the `fuzz_target!` macro... but the other ref-magic works fine :ok_hand:

File: fuzz/Cargo.lock
Comment (wiktor-k) [reply fuzz/Cargo.lock]: Hmm... I wasn't sure if I could have parent project in a normal layout and also fuzz as a workspace member but apparently inserting:

```toml
[workspace]
members = [".", "fuzz"]
```

in the parent Cargo.toml worked :exploding_head:

**Comment by wiktor-k** [comment]
Thank you *very* much for your suggestions @kpcyrd :pray:

**Comment by wiktor-k** [comment]
> However, you can create a new slice with identical pointer and length like so (this is a very cheap operation):

Ha, right. I was aware of the in-memory layout of fat pointers but didn't connect that `&data[..]` creates a new one :man_facepalming: 

> The &mut is needed because it's expected by Message::decode, if it would take something like decode<R: Read>(mut reader: R) -> Result<Self> you could drop the extra &mut too. I'm not sure if that's a worthwhile idea though. Reader is implemented for impl Reader for Decoder<'_> and I'm not sure if the decode function would then insist on taking ownership of Decoder.

Alright. (Un)fortunately I don't control the Decode trait (since it comes from the RustCrypto's crate) and I've come with at different design [for my other SSH crate](https://github.com/wiktor-k/ssh-sig) that uses look-ahead parser (similar to programming languages).

> The patch looks good to me now. 👍

Thanks a lot for your time and explanations. I really appreciate them! :bow: 

I'll let the PR sit for a couple of days in case the other maintainers feel like commenting and if not merge it.

:wave:

**Comment by kpcyrd** [comment]
Sure, you're welcome! :)

**Review by baloo — APPROVED** [review]

**Pull request merged**
